### ISSUE TITLE
Full Server Crash with FFMpeg Producers, throwing GL_INVALID_FRAMEBUFFER_OPERATION_EXT

### ISSUE THREAD

**Issue by LRTNZ** [CasparCG/server#1296]
### Current behaviour

When running Caspar Server on an Ubuntu 18.04 Server machine, with two Quadro P1000 GPUs in it, driving 4 FFpeg producers using NVENC encoding, occasionally it will crash and sit throwing error messages for all eternity until it is stopped.
I wish I could provide a more useful explanation than that, but honestly I do not even know where to begin, other than saying to probably look at the attached log, and fire away with any questions you may have, so we can get to the bottom of this.

### Potentially notable points

- This machine is running headless. This is a requirement for the installation, and I achieved this by configuring virtual displays in the X11 config. Then I am using LXDM to automatically start X11, login to the Caspar user on the machine, and fire off a script to startup the mediascanner and Caspar server in a detached screen.
- We are using two Nvidia P1000s, as they are the only ones that will fit into the server, as it is a 1U unit. One GPU would have been preferable, but we have had to make do.
- We are having to use a workaround to bypass Nvidia's NVENC concurrent session limitation, as it applies across the system, not per card. This is being done using this: https://github.com/keylase/nvidia-patch
- The version of Caspar in use has been built with NVENC enabled using the steps I roughly outlined here: https://github.com/CasparCG/server/issues/1282#issuecomment-620492810
- The server with this on it has been installed into the venue, and is running other systems on it now as well when this problem showed up again. I originally thought it was a one off, having not rebooted the machine after applying the above patch to get around Nvidia's limitations. This does mean that testing on the affected machine will not really be possible.
- I do not have physical access to the machine, as it is in another country, about 3hrs of flying away.

### Steps to reproduce
1. Compile Casparcg for Linux, with NVENC Enabled
2. Install onto an ubuntu machine, with 2x Nvidia GPUS
3. In the config, setup 4x mpegts Multicast streams. Could use this as a channel, changing the GPU index, to split the transcoding equally across the two GPUs:  
```
<channel>
   <video-mode>1080p3000</video-mode>
   <consumers>
       <ffmpeg>
            <path>udp://239.1.1.5:1234?pkt_size=1316</path>
            <args>-codec:v h264_nvenc -profile:v high -cq:v 10 -aud:v 1 -g:v 25 -keyint_min:v 25 -filter:v format=yuv420p -format mpegts</args>
        </ffmpeg>
        </consumers>
</channel>
```
4. Startup the server
5. Startup a video playing on loop on each of these different channels.
6. If one of the channels does not cause a cascade of errors as in the log in the attached zip file, stop the server, and try restarting caspar, and see what happens. Eventually, you will get a spectacular failure like in the attached log file.

### Environment
* Commit: 3294513
* Operating system: Ubuntu Server 18.04
* Intel Xeon Processor
* 2x Nvidia P1000 GPUs, with this patch applied to unlock the NVENC encoding count limit: https://github.com/keylase/nvidia-patch
---

### Full Caspar Log

[caspar_2020-05-12.zip](https://github.com/CasparCG/server/files/4615242/caspar_2020-05-12.zip)

**Comment by hreinnbeck** [comment]
Does this happen if you have all the 4 channels only use the second P1000?

**Comment by walterav1984** [comment]
If its running headless are you sure no screensaver is running after a while?

**Comment by LRTNZ** [comment]
There is no desktop GUI installed or running, so no. This also does not always happen, it will randomly show up not long after startup if it is going to happen, as before this, the same install and settings had been running flawlessly nearly 24/7 for about a week and a half before the server had to be shut down to be moved to it's installation site.

**Comment by LRTNZ** [comment]
> Does this happen if you have all the 4 channels only use the second P1000?

Well, as we are live transcoding 4 1920x1080p 24fps streams, I don't particularly want to be putting all 4 streams onto one GPU. Also, as I have said, it is periodically that it will randomly do it, often after it is just started. Because if this doesn't happen at startup, Caspar will keep running forever without an issue.

Did you see the log I attached out of curiosity?

**Comment by hreinnbeck** [comment]
Yeah I read the log, that's why I wanted to see if it would go away with the nvenc sessions on the second GPU.  I do 14x 1080p50 on a P2000 so I'd think the P1000 should easily handle the 4 streams.  Does nvidia-smi give any indication of the GPUs being over utilized?

**Comment by LRTNZ** [comment]
14x streams? What arguments are you using for that?
Also, is that on Linux?
May have to see if I can get them working on one GPU. Nvidia smi seems to have always shown these GPUs to have headroom to spare whenever I have checked that. 

The P2200 is the lowest NVIDIA GPU that 'natively' has the encoding session count unlocked, so maybe that could be a factor?

However this all still leaves the question - why does it work ~90% of the time, and if it is going to fail, it is just after a few media plays, never days, or weeks after Caspar started running.

**Comment by Julusian** [comment]
This sounds like it could be similar related to the screen consumer sometimes filing to open at startup on Linux.

Perhaps there is some race condition being hit at startup causing one of the channels to fail to initialise? That is not something it would ever recover from.

I need to do some testing myself, but I suspect that the key factor here will be Linux, and using nvenc could well be irrelevant

**Comment by LRTNZ** [comment]
> This sounds like it could be similar related to the screen consumer sometimes filing to open at startup on Linux.
> 
> I need to do some testing myself, but I suspect that the key factor here will be Linux, and using nvenc could well be irrelevant 

So, would using 2x GPUs have anything to do with it? The channel I provided above is along the lines of the actual config being used (the one I have had an issue with at the customers site has tweaks in the audio encoding), however I have had it die on the them in the past as well.

**Comment by LRTNZ** [comment]
> Perhaps there is some race condition being hit at startup causing one of the channels to fail to initialise? That is not something it would ever recover from.

Is there a way to stop this? As the odd thing is that all 4 channels can and will playout media before it will suddenly happen on the rare occasion it does.

**Comment by Julusian** [comment]
I did have a quick test the other day, and was not able to cause a crash like this. 4 channels doing nvenc streams (didnt view them, but I assume it was actually doing it). This was on a single P4000, so perhaps the 2 gpu is a factor but I couldnt explain why.

> Is there a way to stop this? As the odd thing is that all 4 channels can and will playout media before it will suddenly happen on the rare occasion it does.

Do you mean that it can happen a few minutes after every channel is playing, or does it happen when one of the channels starts playing for the first time (and the others may be doing something)?

**Comment by LRTNZ** [comment]
> I did have a quick test the other day, and was not able to cause a crash like this. 4 channels doing nvenc streams (didnt view them, but I assume it was actually doing it). This was on a single P4000, so perhaps the 2 gpu is a factor but I couldnt explain why.
>
I wonder (And this is a stab in the dark), as to whether or not this is something to do with the mixer and what GPU it is on, relevant to the channel? Then again, the issue, when it occurs, only occurs once you start media playback, on any of the channels. 

Thinking about it some more, I don't recall seeing this error before adding the second GPU? But that shouldn't make a difference, unless there is a chance OpenGL randomly grabs the wrong GPU at start-up? Because it is ffmpeg handling the usage of the second GPU, purely for nvenc encoding.

> Do you mean that it can happen a few minutes after every channel is playing, or does it happen when one of the channels starts playing for the first time (and the others may be doing something)?
> 
I have had it happen after the first loop of playing a clip, when starting a clip to play ( with or without looping, or any other options), different formats of videos, maybe after playing two clips, etc. Though almost always if it will happen, it will be on the first media playback on the server after starting it up.
